python_modules/dagster/dagster/_core/pipes/utils.py

```diff
@@ -427,11 +476,64 @@ def _logs_thread(
             )
             raise
         finally:
-            for log_reader in self.log_readers:
+            for log_reader in self.log_readers.values():
                 if log_reader.is_running():
                     log_reader.stop()
 
 
+class PipesBlobStoreMessageReader(PipesThreadedMessageReader):
+    """Message reader that reads a sequence of message chunks written by an external process into a
+    blob store such as S3, Azure blob storage, or GCS.
+
+    The reader maintains a counter, starting at 1, that is synchronized with a message writer in
+    some pipes process. The reader starts a thread that periodically attempts to read a chunk
+    indexed by the counter at some location expected to be written by the pipes process. The chunk
+    should be a file with each line corresponding to a JSON-encoded pipes message. When a chunk is
+    successfully read, the messages are processed and the counter is incremented. The
+    :py:class:`PipesBlobStoreMessageWriter` on the other end is expected to similarly increment a
+    counter (starting from 1) on successful write, keeping counters on the read and write end in
+    sync.
+
+    If `log_readers` is passed, the message reader will start the passed log readers when the
+    `opened` message is received from the external process.
+
+    Args:
+        interval (float): interval in seconds between attempts to download a chunk
+        log_readers (Optional[Mapping[str, PipesLogReader]]): A mapping of arbitrary names to readers for logs.
+    """
+
+    counter: int
+
+    def __init__(
+        self,
+        interval: float = 10,
+        log_readers: Optional[Sequence["PipesLogReader"]] = None,
+    ):
+        super().__init__(interval=interval, log_readers=log_readers)
+
+        self.counter = 1
+
+    @abstractmethod
+    def download_messages_chunk(self, index: int, params: PipesParams) -> Optional[str]:
+        ...
+        # historical reasons, keeping the original interface of PipesBlobStoreMessageReader
+
+    def download_messages_parts(
+        self, cursor: Any, params: PipesParams
```

Comment:
cursor: Any should be `cursor: TCursor` right

Reply:
Also please add more of an explanatory comment about `download_messages_chunk` and `download_messages_parts`. Very confusing for anyone fresh to the code to see both methods here.